### DISCUSSION TITLE
Output the AST as JSON

### DISCUSSION THREAD

**Discussion by charles545587** [kivikakk/comrak#310] [Q&A]
Hi,

I would like to take a markdown file, generate the AST and then output the AST as JSON object/file so that I can then perform a further action on it. I am still exploring the options, but is it possible to just extract the AST without parsing it into a different format?

**Comment by kivikakk** [comment]
Hi there! There are two main options that present themselves.

This isn't currently (directly) possible using just the binary itself. The closest thing it has to "output a raw AST" is the XML output mode — which is actually pretty damn close! You could post-process the XML into JSON and you're likely to have the fidelity you want.

If you want more control than that, or if the XML output as it stands is missing some pieces you need from the AST nodes, you'd need to write something yourself that uses Comrak as a library to parse the input, and then output the JSON you want directly. There's a (very general) [`s-expr` example](https://github.com/kivikakk/comrak/blob/3e4c0936a372cbe2c6840bcdbee83ed5013dc5c5/examples/s-expr.rs) included that dumps a Markdown file as generic S-expressions; you might find this example helpful as a basis for your own tool.

**Reply by charles545587** [reply]
Thanks for the swift response, I will try both approaches and see how I get on.
